scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -201,108 +247,221 @@ def _highly_variable_genes_single_batch(
-        ) / disp_std_bin[df["mean_bin"].to_numpy()].to_numpy()
+        disp_stats = _stats_seurat(df["mean_bin"], disp_grouped)
     elif flavor == "cell_ranger":
-        from statsmodels import robust
+        disp_stats = _stats_cell_ranger(df["mean_bin"], disp_grouped)
+    else:
+        raise ValueError('`flavor` needs to be "seurat" or "cell_ranger"')
 
-        df["mean_bin"] = pd.cut(
-            df["means"],
-            np.r_[-np.inf, np.percentile(df["means"], np.arange(10, 105, 5)), np.inf],
-        )
-        disp_grouped = df.groupby("mean_bin", observed=True)["dispersions"]
-        disp_median_bin = disp_grouped.median()
-        # the next line raises the warning: "Mean of empty slice"
-        with warnings.catch_warnings():
-            warnings.simplefilter("ignore")
-            disp_mad_bin = disp_grouped.apply(robust.mad)
-            df["dispersions_norm"] = (
-                df["dispersions"].to_numpy()
-                - disp_median_bin[df["mean_bin"].to_numpy()].to_numpy()
-            ) / disp_mad_bin[df["mean_bin"].to_numpy()].to_numpy()
+    # actually do the normalization
+    df["dispersions_norm"] = (df["dispersions"] - disp_stats["avg"]) / disp_stats["dev"]
+    df["highly_variable"] = _subset_genes(
+        adata,
+        mean=mean,
+        dispersion_norm=df["dispersions_norm"].to_numpy(),
+        cutoff=cutoff,
+    )
+
+    df.set_index(adata.var_names, inplace=True)
+    df.index.name = "gene"
+    return df
+
+
+def _get_mean_bins(
+    means: pd.Series, flavor: Literal["seurat", "cell_ranger"], n_bins: int
+) -> pd.Series:
+    if flavor == "seurat":
+        bins = n_bins
+    elif flavor == "cell_ranger":
+        bins = np.r_[-np.inf, np.percentile(means, np.arange(10, 105, 5)), np.inf]
     else:
         raise ValueError('`flavor` needs to be "seurat" or "cell_ranger"')
-    dispersion_norm = df["dispersions_norm"].to_numpy()
-    if n_top_genes is not None:
-        dispersion_norm = dispersion_norm[~np.isnan(dispersion_norm)]
-        dispersion_norm[
-            ::-1
-        ].sort()  # interestingly, np.argpartition is slightly slower
-        if n_top_genes > adata.n_vars:
-            logg.info("`n_top_genes` > `adata.n_var`, returning all genes.")
-            n_top_genes = adata.n_vars
-        if n_top_genes > dispersion_norm.size:
-            warnings.warn(
-                "`n_top_genes` > number of normalized dispersions, returning all genes with normalized dispersions.",
-                UserWarning,
-            )
-            n_top_genes = dispersion_norm.size
-        disp_cut_off = dispersion_norm[n_top_genes - 1]
-        gene_subset = np.nan_to_num(df["dispersions_norm"].to_numpy()) >= disp_cut_off
+
+    return pd.cut(means, bins=bins)
+
+
+def _stats_seurat(mean_bins: pd.Series, disp_grouped: SeriesGroupBy) -> pd.DataFrame:
+    """Compute mean and std dev per bin."""
+    disp_bin_stats = disp_grouped.agg(avg="mean", dev=partial(np.std, ddof=1))
```

Comment:
```suggestion
    disp_bin_stats = disp_grouped.agg(avg="mean", dev="std")
```

This should be equivalent and use the pandas method, which could be optimized by pandas. Maybe faster (but I'm not 100% convinced) in a quick check:

```python
In [16]: %timeit res1 = df.groupby("cat").agg(["mean", "std"])
1.48 ms ± 46.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [14]: %timeit res2 = df.groupby("cat").agg(["mean", partial(np.std, ddof=1)])
1.77 ms ± 56.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Reply:
I didn‘t realize that pandas has a different default. Sure!